platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt

```diff
@@ -0,0 +1,125 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.problems.internal
+
+import org.gradle.api.internal.DocumentationRegistry
+import org.gradle.api.internal.StartParameterInternal
+import org.gradle.api.internal.file.temp.TemporaryFileProvider
+import org.gradle.api.logging.Logger
+import org.gradle.api.logging.Logging
+import org.gradle.api.problems.internal.FileLocation
+import org.gradle.api.problems.internal.Problem
+import org.gradle.internal.buildoption.InternalOptions
+import org.gradle.internal.cc.impl.problems.BuildNameProvider
+import org.gradle.internal.cc.impl.problems.JsonSource
+import org.gradle.internal.cc.impl.problems.JsonWriter
+import org.gradle.internal.concurrent.ExecutorFactory
+import org.gradle.internal.configuration.problems.CommonReport
+import org.gradle.internal.configuration.problems.FailureDecorator
+import org.gradle.internal.configuration.problems.StructuredMessage
+import org.gradle.internal.configuration.problems.writeError
+import org.gradle.internal.configuration.problems.writeStructuredMessage
+import org.gradle.internal.logging.ConsoleRenderer
+import org.gradle.internal.operations.OperationIdentifier
+import org.gradle.internal.problems.failure.FailureFactory
+import org.gradle.problems.buildtree.ProblemReporter
+import java.io.File
+
+val logger: Logger = Logging.getLogger(DefaultProblemsReportCreator::class.java)
+
+class DefaultProblemsReportCreator(
+    executorFactory: ExecutorFactory?,
+    temporaryFileProvider: TemporaryFileProvider?,
+    internalOptions: InternalOptions?,
+    startParameter: StartParameterInternal,
+    private val failureFactory: FailureFactory,
+    private val buildNameProvider: BuildNameProvider
+) : ProblemReportCreator {
+
+    private val report = CommonReport(executorFactory!!, temporaryFileProvider!!, internalOptions!!, "problem-report")
+    private val taskNames: List<String> = startParameter.taskNames
+    private var problemCount = 0
+
+    private val failureDecorator = FailureDecorator()
+
+    override fun getId(): String {
+        return "DefaultProblemsReportCreator"
+    }
+
+    override fun report(reportDir: File, validationFailures: ProblemReporter.ProblemConsumer) {
+        report.writeReportFileTo(reportDir, object : JsonSource {
+            override fun writeToJson(jsonWriter: JsonWriter) {
+                with(jsonWriter) {
+                    property("reportContext", "problems report")
```

Comment:
❌  Having this marker implies that the extra data expected by the problems report will be at the top level of the JSON as well. I think it makes it harder to maintain and review.

Instead of mixing the data, let's introduce a single top-level field in the JSON that will carry all the extra data needed for the Problems Report. The presence of that field will indicate to the JS report logic that it's a Problems Report rather than CC report, and it can proceed with processing this data.

Reply:
I'm all for it, but can we do that later when we ironed out what we exactly want to transfer into the json?
The purpose for now is the separation, so we can do different things for CC and problems.